### PR TITLE
NAS-125303 / 24.04 / Differentiate between datasets and directories in file picker

### PR DESCRIPTION
**Pull request by bvasilenko**
**Testing**

1. Go to Shell and create a few directories under `/mnt/<your pool name>`
2.  On Sharing page, in Windows SMB shares, click **Add**
3. In the **Add SMB** form, at Basic block you can see a File/Directory picker
**Expected:** directory and dataset have 2 different icons

### CODE REVIEW

**Review by dszidi — APPROVED** [review]
Seems to be working correctly and the new directory icon matches our designs.
LGTM

image


image

**Review by undsoft — CHANGES_REQUESTED** [review]
`a`, `b` and `a-lot-of-datasets` appear to be datasets, not folders as shown in screenshot.

**Comment by bvasilenko** [comment]
@undsoft @dszidi would you please provide output of these 3 commands 

```
call "filesystem.listdir" "/mnt/APPS/APPS/a"
call "filesystem.listdir" "/mnt/APPS/APPS/b"
call "filesystem.listdir" "/mnt/APPS/APPS/a-lot-of-datasets"
```

from the machine from where your screenshot is taken from?

**Comment by bvasilenko** [comment]
@undsoft in that particular case it’s impossible to distinguish this particular dataset from a directory. 

Taking into account that
- this is rare edge case (replicated read-only file system)
- the problem exists in the middleware query `pool.dataset.query` which does not return these read-only replicated datasets in the list of datasets
- common user's expectations are that icon is rather cosmetical hint then a strict indication of whether this a directory or a dataset

I'd suggested to proceed with the merge

**Review by undsoft — CHANGES_REQUESTED** [review]
Mixing icons is confusing. We shouldn't do it.
Let's push middleware to try to solve this, if not we can discuss whether this is something we can solve on frontend.
The alternative is not to have this feature at all – it's better not to confuse the user.

File: src/app/services/filesystem.service.ts [hunk 1]
```diff
@@ -20,24 +20,30 @@ export class FilesystemService {
...
-        [node.data.path, typeFilter, { order_by: ['name'], limit: 1000 }],
-      ).pipe(
-        map((files) => {
+      if (options.hideZfs) {
+        typeFilter.push(['is_ctldir', '=', false]);
+      }
+
+      return forkJoin([
+        this.ws.call('filesystem.listdir', [node.data.path, typeFilter, { order_by: ['name'], limit: 1000 }]),
+        this.ws.call('pool.dataset.query', [[], { select: ['mountpoint'] }]),
```
[the remark this answers is not in the document]
Comment (anodos325) [reply src/app/services/filesystem.service.ts:44]: The filesystem.listdir output is correct. If the dataset is not mounted, then that component is not a dataset. It is a directory.

**Review by undsoft — CHANGES_REQUESTED** [review]
Okay, I've tested on m40. The old dataset is gone, so I cannot reproduce the original issue. It seems to work fine with old version (where it didn't query for datasets).
Let's see if we can incorporate https://ixsystems.atlassian.net/browse/NAS-125555?atlOrigin=eyJpIjoiMWY0YmUzMGEzZmI0NGQ4ZmI1ZjY0MmQ2NTJkZDFhNmUiLCJwIjoiaiJ9 in some way.

**Comment by bvasilenko** [comment]
> Let's see if we can incorporate https://ixsystems.atlassian.net/browse/NAS-125555?atlOrigin=eyJpIjoiMWY0YmUzMGEzZmI0NGQ4ZmI1ZjY0MmQ2NTJkZDFhNmUiLCJwIjoiaiJ9 in some way.

@undsoft this is already done at commit https://github.com/truenas/webui/pull/9254/commits/2781a6debd1efe2457657c58e229033104289514#diff-bc2d1338cf9a61ec3d6f90e024f22dc0a955ee60106b3c1a5a1b93e0f9b60c3dR44

So I suppose it's ready to be merged

**Comment by bvasilenko** [comment]
encrypted directory with "lock" icon:

image

**Review by undsoft — APPROVED** [review]
Looks good, thanks.

**Pull request merged**